titus-server-master/src/main/java/com/netflix/titus/master/clusteroperations/ClusterAgentAutoScaler.java

```diff
@@ -473,9 +474,13 @@ private Completable createResetOverrideStatusesCompletable(List<AgentInstance> r
         return v3JobOperations.getTasks().stream().collect(Collectors.toMap(Task::getId, Function.identity()));
     }
 
-    private Map<Tier, Set<String>> getFailedTaskIds(Map<FailureKind, Map<String, List<TaskPlacementFailure>>> taskPlacementFailures, Set<FailureKind> ignoring) {
+    /**
+     * @param <T> generic helper "trick" to capture the wildcard. See https://docs.oracle.com/javase/tutorial/java/generics/capture.html
+     */
+    private <T> Map<Tier, Set<String>> getFailedTaskIdsByTier(Map<FailureKind, Map<T, List<TaskPlacementFailure>>> taskPlacementFailures,
+                                                              Set<FailureKind> ignoring) {
         Map<Tier, Set<String>> failedTaskIdsByTier = new HashMap<>();
-        for (Map<String, List<TaskPlacementFailure>> failuresByTaskId : taskPlacementFailures.values()) {
+        for (Map<T, List<TaskPlacementFailure>> failuresByTaskId : taskPlacementFailures.values()) {
```

Comment:
How the keys are compared here? By object ref or value?

Reply:
object ref, we have some other code indexing maps by `TaskRequests` (and Fenzo does it too internally). Since these are mutable objects that we keep references around (forever during its lifecycle), I think this is fine.